definitions/npm/knex_v0.12.x/flow_v0.38.x-/knex_v0.12.x.js

```diff
@@ -110,23 +117,24 @@ declare class Knex$QueryBuilder mixins Promise {
   pluck(column: string): this,
   first(): this,
   from(table: string): this,
-  from(builder: Knex$QueryBuilderFn | Knex$Knex | Knex$QueryBuilder): this,
-
-  insert(val: Object | Object[]): this,
+  from(builder: Knex$QueryBuilderFn<T> | Knex$Knex<T> | Knex$QueryBuilder<T>): this,
+  insert(val: Object | Object[], returning?: string|Array<string>): this,
   del(): this,
   delete(): this,
   update(column: string, value: any): this,
-  update(val: Object): this,
+  update(val: Object, returning?: string|Array<string>): this,
   returning(columns: string[]): this
 }
 
-declare class Knex$Knex mixins Knex$QueryBuilder, Promise, events$EventEmitter {
-  static (config: Knex$Config): Knex$Knex,
+declare class Knex$Knex<T> mixins Knex$QueryBuilder<T>, Promise, events$EventEmitter {
+  static (config: Knex$Config): Knex$Knex<T>,
```

Comment:
Wow, this all seems *very* tricky.  There are at least a couple things I think we need to consider:

1. We don't want to over-constrain the type parameter of `Knex$Knex`, since we'd like it to represent the whole database, which could contain many tables (yelp-gil-service's db is an outlier with only 1 table).
2. Joins seems especially tricky, since they are a single operation, e.g. `innerJoin` that involves at least 2 types--one for each table involved in the join.  They may even involve *3 types*--one for each table, plus one for the result type if the two tables are joined in a projection (i.e. a subset of fields from each table).  Sometimes that 3rd type might be a proper intersection of the two other types, but not always.

I have some ideas, but nothing concrete yet ... let's talk about it tomorrow if we have time.